### PR TITLE
[Merged by Bors] - feat(Mathlib/RingTheory/Ideal/Operations): simplify definition of radical

### PR DESCRIPTION
**Pull request by AntoineChambert-Loir**
* `add_pow_mem_of_pow_mem_of_le` says that (a + b) ^ k belongs to an ideal I if a ^ m and b ^ n belong to that ideal, and `m + n ≤ k + 1`.
* `Commute.add_pow_of_add_le_succ_eq_zero_of_pow_eq_zero` says that `(a + b) ^ k = 0` if `a ^ m = 0`, `b^ n = 0`, and `m + n ≤ k + 1`.
* this is used to simplify the definition of docs#Ideal.radical and the definition of docs#Commute.isNilpotent_add

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/RingTheory/Ideal/Basic.lean [hunk 1]
```diff
@@ -578,6 +578,38 @@ theorem pow_mem_of_mem (ha : a ∈ I) (n : ℕ) (hn : 0 < n) : a ^ n ∈ I :=
...
+  rw [add_pow]
+  apply I.sum_mem
+  intro c _
+  apply mul_mem_right
+  by_cases h : m ≤ c
+  · exact I.mul_mem_right _ (I.pow_mem_of_pow_mem ha h)
+  · refine I.mul_mem_left _ (I.pow_mem_of_pow_mem hb ?_)
+    simp only [not_le] at h
+    rw [add_comm, Nat.add_sub_assoc (le_of_lt h)]
+    apply Nat.le_add_right
```
Comment (ocfnash) [inline Mathlib/RingTheory/Ideal/Basic.lean:597]: Optional:
```suggestion
  rw [Commute.add_pow' (Commute.all _ _)]
  refine I.sum_mem fun ⟨i, j⟩ hij ↦ ?_
  rw [nsmul_eq_mul]
  apply mul_mem_left
  by_cases h : m ≤ i
  · exact I.mul_mem_right _ (I.pow_mem_of_pow_mem ha h)
  · refine I.mul_mem_left _ (I.pow_mem_of_pow_mem hb ?_)
    linarith [Finset.mem_antidiagonal.mp hij]
```

Comment (ocfnash) [reply Mathlib/RingTheory/Ideal/Basic.lean:597]: (and similarly for `add_pow_add_succ_mem_of_pow_succ_mem` below)

**Review by ocfnash — APPROVED** [review]
Thanks, one optional suggestion.

bors d+

**Comment by AntoineChambert-Loir** [comment]
Actually, I just realized (ON made me…) that docs#Commute.isNilpotent_add merits being adjusted as well. I need a few hours…

File: Mathlib/RingTheory/Ideal/Basic.lean [hunk 2]
```diff
@@ -578,6 +578,24 @@ theorem pow_mem_of_mem (ha : a ∈ I) (n : ℕ) (hn : 0 < n) : a ^ n ∈ I :=
...
+  apply I.sum_mem
+  intro c _
+  apply mul_mem_right
+  by_cases h : m ≤ c
+  · exact I.mul_mem_right _ (I.pow_mem_of_pow_mem ha h)
+  · refine I.mul_mem_left _ (I.pow_mem_of_pow_mem hb ?_)
+    simp only [not_le, Nat.lt_iff_add_one_le] at h
+    rw [Nat.sub_sub, add_comm 1, add_comm m, Nat.add_sub_assoc h]
+    apply Nat.le_add_right
+
```
Comment (ocfnash) [inline Mathlib/RingTheory/Ideal/Basic.lean:598]: How about:
```suggestion
variable {I}

theorem add_pow_mem_of_pow_mem_of_le {m n k : ℕ}
    (ha : a ^ m ∈ I) (hb : b ^ n ∈ I) (hk : m + n ≤ k + 1) :
    (a + b) ^ k ∈ I := by
  rw [(Commute.all _ _).add_pow']
  refine I.sum_mem fun ⟨i, j⟩ hij ↦ ?_
  rw [nsmul_eq_mul]
  apply mul_mem_left
  by_cases h : m ≤ i
  · exact I.mul_mem_right _ (I.pow_mem_of_pow_mem ha h)
  · refine I.mul_mem_left _ (I.pow_mem_of_pow_mem hb ?_)
    linarith [Finset.mem_antidiagonal.mp hij]

theorem add_pow_add_pred_mem_of_pow_mem  {m n : ℕ}
    (ha : a ^ m ∈ I) (hb : b ^ n ∈ I) :
    (a + b) ^ (m + n - 1) ∈ I :=
  add_pow_mem_of_pow_mem_of_le ha hb <| by rw [← Nat.sub_le_iff_le_add]

```

Comment (AntoineChambert-Loir) [reply Mathlib/RingTheory/Ideal/Basic.lean:598]: I'll do that, as well as for the nilpotent version.

File: Mathlib/RingTheory/Nilpotent.lean [hunk 3]
```diff
@@ -311,17 +311,27 @@ section Semiring
 
 variable [Semiring R] (h_comm : Commute x y)
 
-theorem isNilpotent_add (hx : IsNilpotent x) (hy : IsNilpotent y) : IsNilpotent (x + y) := by
-  obtain ⟨n, hn⟩ := hx
-  obtain ⟨m, hm⟩ := hy
-  use n + m - 1
+theorem add_pow_of_add_le_succ_eq_zero_of_pow_eq_zero {m n k : ℕ}
```
Comment (ocfnash) [inline Mathlib/RingTheory/Nilpotent.lean:314]:
```suggestion
theorem add_pow_eq_zero_of_add_le_succ_of_pow_eq_zero {m n k : ℕ}
```

File: Mathlib/RingTheory/Ideal/Basic.lean [hunk 4]
```diff
@@ -578,6 +578,31 @@ theorem pow_mem_of_mem (ha : a ∈ I) (n : ℕ) (hn : 0 < n) : a ^ n ∈ I :=
...
+  intro c _
+  apply mul_mem_right
+  by_cases h : m ≤ c
+  · exact I.mul_mem_right _ (I.pow_mem_of_pow_mem ha h)
+  · refine I.mul_mem_left _ (I.pow_mem_of_pow_mem hb ?_)
+    simp only [not_le, Nat.lt_iff_add_one_le] at h
+    rw [Nat.le_sub_iff_add_le, ← add_le_add_iff_right 1]
+    exact le_trans (by rwa [add_comm _ n, add_assoc, add_le_add_iff_left]) hk
+    rw [← add_le_add_iff_right 1]
+    exact le_trans h (le_trans (Nat.le_add_right _ _) hk)
```
Comment (ocfnash) [inline Mathlib/RingTheory/Ideal/Basic.lean:599]: I can't immediately find it in the style guide but we prefer not to spawn subgoals during a rewrite like this. (In particular, any code that contains and `exact` statement which is followed by more code at the same indentation level is almost always not our preferred style.)
```suggestion
    have hck : c ≤ k := by
      rw [← add_le_add_iff_right 1]
      exact le_trans h (le_trans (Nat.le_add_right _ _) hk)
    rw [Nat.le_sub_iff_add_le hck, ← add_le_add_iff_right 1]
    exact le_trans (by rwa [add_comm _ n, add_assoc, add_le_add_iff_left]) hk
```

File: Mathlib/RingTheory/Nilpotent.lean (hunk 3 above)
Comment (ocfnash) [reply Mathlib/RingTheory/Nilpotent.lean:314]: after which you'll need to adjust where it is used below

**Review by ocfnash — APPROVED** [review]
Thanks for sticking with this!

bors d+

**Pull request closed without merging**